plugins/module_utils/module_helper.py

```diff
@@ -151,23 +173,26 @@ def __init__(self, module=None):
         if module:
             self.module = module
 
-        if not isinstance(module, AnsibleModule):
+        if isinstance(self.module, dict):
+            if self.named_deprecations:
```

Comment:
Some validation of `self.named_deprecations` would be good, i.e. to make sure that all required fields are in there. The problem with doing deprecations in this general way is that it disables all deprecation validations (and there are several important ones). So either you have to redo everything here (not advisable), or you need to revise the mechanism so that the user-code calls `module.deprecate()` itself (and gets the message formatted by the helper code). The latter is probably the best, to make sure everything is correctly validated.

Reply:
not sure if I got it... I thought I was doing just that in the trigger_named_deprecation() method below. See line 264.

Reply:
The problem is that you are not passing explicit values in the function call, so the static analysis the pylint plugin does won't work. It checks the explicitly specified strings `version`, `date`, `collection_name` to the `module.deprecate()` call. If they are not specified explicitly, there's nothing the analysis can do.

So it's necessary that the module itself calls `module.deprecate()` and supplies these values.

Reply:
See here for the pylint plugin: https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py#L204-L254

Reply:
What a bummer! But I see your point - and I would not like to go around the validation. On it.

Reply:
You could ask the module to pass something like `lambda module, msg: module.deprecate(msg, version='4.0.0', removed_from_collection='community.general')` (or without the `module` part, depending on how exactly it will be used) to the function instead of a kwargs. Then static analysis will still work (since it doesn't care about the message, only about the remainder).

Reply:
This is the first time I am seeing pylint's internal organs dissected, but if I got it right, *any* function call named `deprecate()` will be analysed by this checker you sent me the link. If that's so, then we could trick pylint with:
```python
    @staticmethod
    def deprecate(msg, version, date, collection):
        return (msg, version, date, collection)
```
at the helper, then the implementing module would declare:
```python
    named_deprecations = {
        'bla': ModuleHelper.deprecate("blablabla", version="4.0.0", collection_name="c.g")
    }
```
could we not?

Reply:
It's dirty and silly, but I think it beats forcing the user to write a lambda for each one of them.

Reply:
Come to think of it, this semantically a bad idea because it reads as if we were _effectively_ deprecating something when calling that. Maybe a helper class, say `NamedDeprecation`, to indicate this is something else.

Reply:
You might want to check whether it really also accepts a call on a class (and not just on a class instance). (Not sure whether pylint really knows the difference though...)

Reply:
@felixfontein it does accept on a class, but I tried getting something like:
```python
    n_d = (
        MH.NamedDeprecations('facts').deprecate('some message', version='1.1.1', community_name='c.g')
```
But that is not picked up by pylint, so I had the static method put in place.

Reply:
I have used xfconf as a testbed for it, it ended up like this:
```python
    named_deprecations = dict(
        facts=CmdStateModuleHelper.PrepareNamedDeprecation.deprecate(
            'Module use of facts is not a best practice and will be deprecated. Use return values instead.',
            version='3.0.0', collection_name='community.general'),
    )
```